alphacsc/_d_solver.py

```diff
@@ -194,13 +194,13 @@ def prox(D, step_size=None):
 
         return prox
 
-    def _get_grad(self, z_encoder):
+    def _get_grad(self, X, z_hat, constants):
```

Comment:
something like this?
```suggestion
    def _get_grad(self, X, z_encoder):
        constants = z_encoder.get_constants()
        z_hat = z_encoder.get_z_hat() if self.loss == "l2" else None
```

Reply:
At least we should only call `get_z_hat` here, not in `update_D` and follow up `update_u/v`.